[WOO13-84] fix: address maintainability issues with v4 implementation of shipping endpoints

ayushpahwa opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR refactors the V4 REST API shipping endpoints to use a service layer pattern. The main goal is to move business logic out of public core classes (`WC_Shipping_Method`, `WC_Shipping_Zone`) into an internal service.

**Key architectural improvements:**

1. New `ShippingService` class (`src/Internal/Shipping/ShippingService.php`). Handles all shipping zone and method business logic
2. Controller Simplification to focus on HTTP layer (validation, status codes, response formatting)
3. Optimise DB calls by combining `enabled` and `order` updates into single `UPDATE` query
4. Remove business logic previously added in PR #61124 from `WC_Shipping_Method` or `WC_Shipping_Zone`

Closes #WOO13-84

### Screenshots or screen recordings:

_No UI changes - this is an internal API refactoring._

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
- Install and activate WooCommerce
- Ensure shipping is enabled in WooCommerce settings
- Have API authentication set up (Application Passwords or OAuth)
- Enable REST API v4 feature flag:
    - Ensure the WC Beta Tester plugin is installed. 
    - Navigate to `/wp-admin/tools.php?page=woocommerce-admin-test-helper&tab=features` and enable the flag for `rest-api-v4`
- Authorization
    - Create an application password using `wp-env run cli wp user application-password create admin "dev" --porcelain`
    - In your preferred app, or using cURL, get a zone ID from the list endpoint: `curl -s -u admin:<pass> '<host>/wp-json/wc/v4/shipping-zones'`
    - Fetch zone details: `curl -s -u admin:<pass> '<host>/wp-json/wc/v4/shipping-zones/{id}'`
    - Note down the zone ID and the list of the methods
- Use the [Bruno collection](https://github.com/user-attachments/files/23245837/Shipping.Service.APIs.json) to run the tests mentioned below


#### Test 1: Create Shipping Zone with Method

1. Send POST request to `/wp-json/wc/v4/shipping-zones`:
   ```json
   {
     "name": "US Zone",
     "locations": [
       {
         "code": "US",
         "type": "country"
       }
     ]
   }
   ```
2. **Expected**: Zone created successfully with 201 status
3. Note the `id` from response

5. Send POST request to `/wp-json/wc/v4/shipping-zone-method`:
   ```json
   {
     "zone_id": <zone_id_from_step_3>,
     "method_id": "flat_rate",
     "enabled": false,
     "order": 1,
     "settings": {
       "title": "Standard Shipping",
       "cost": "10.00"
     }
   }
   ```
6. **Expected**: Method created successfully with 201 status
7. **Expected**: Response shows `enabled: false` (not `true`)
8. **Expected**: Settings contain the provided values
9. Make a get call to `/wp-json/wc/v4/shipping-zones` to ensure the zone is created.

#### Test 2: Update Shipping Method

1. Using the method from Test 1, send PUT request to `/wp-json/wc/v4/shipping-zone-method/<instance_id>`:
   ```json
   {
     "enabled": true,
     "order": 5,
     "settings": {
       "title": "Updated Shipping",
       "cost": "15.00"
     }
   }
   ```
2. **Expected**: Method updated successfully with 200 status
3. **Expected**: Response reflects all updated values
4. **Expected**: Settings are sanitized (no XSS if you try to inject scripts)

#### Test 3: Update Zone Details

1. Send PUT request to `/wp-json/wc/v4/shipping-zones/<zone_id>`:
   ```json
   {
     "name": "United States Shipping",
     "order": 10,
     "locations": [
       {
         "code": "US:CA",
         "type": "state"
       },
       {
         "code": "US:NY",
         "type": "state"
       }
     ]
   }
   ```
2. **Expected**: Zone updated successfully
3. **Expected**: Locations are properly formatted in response

#### Test 4: Rest of World Zone Protection

1. Send PUT request to `/wp-json/wc/v4/shipping-zones/0`:
   ```json
   {
     "name": "Renamed Rest of World"
   }
   ```
2. **Expected**: 400 Bad Request
3. **Expected**: Error message: "Cannot change name of 'Rest of the World' zone."

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ayushpahwa committed:
create: init new service for handling shipping operations

ayushpahwa committed:
update: use the service to handle zone fetch

ayushpahwa committed:
update: move shipping zone update method to service

ayushpahwa committed:
update: use the service methods to create and udpate the zones

ayushpahwa committed:
chore: add changlog entry

ayushpahwa committed:
delete: remove business logic code from the data models

ayushpahwa committed:
update: add new update methods in shipping service class

ayushpahwa committed:
update: use the new methods from service in the controller

ayushpahwa committed:
chore: fix lint issues

ayushpahwa committed:
test: add coverage for new service

ayushpahwa committed:
test: move service related tests to the correct file

ayushpahwa committed:
chore: fix lint issues

ayushpahwa committed:
chore: fix lint issues

ayushpahwa committed:
test: use correct method for schema tests

ayushpahwa commented on plugins/woocommerce/src/Internal/Shipping/ShippingService.php:
```diff
@@ -0,0 +1,285 @@
+		);
+
+		return $zones;
+	}
```
rest of the world can not have locations, that is the point, right?

ayushpahwa replied:
Good catch, I have simplified this now.

ayushpahwa committed:
update: fix the param for the filter

ayushpahwa committed:
chore: remove workaround for rest of the world id handling

ayushpahwa commented on plugins/woocommerce/src/Internal/Shipping/ShippingService.php:
```diff
@@ -0,0 +1,285 @@
+		 * @param WC_Shipping_Method   $this              Shipping method instance.
+		 */
+		$filtered_settings = apply_filters( 'woocommerce_shipping_' . $method->id . '_instance_settings_values', $instance_settings, $this );
+		$result            = update_option( $method->get_instance_option_key(), $filtered_settings );
```
That's a great catch, I have updated it now.

ObliviousHarmony commented on plugins/woocommerce/includes/abstracts/abstract-wc-shipping-method.php:
Since you added this method too, how do you feel about removing it? `$enabled` is a public variable and, as far as I can tell, is normally just set directly.

ayushpahwa replied:
Yup, makes sense. Updated

ObliviousHarmony commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/ShippingZoneMethod/Controller.php:
```diff
@@ -138,14 +147,17 @@ public function create_item( $request ) {
 		}
 
+		// Reload the data.
+		$method = WC_Shipping_Zones::get_shipping_method( $instance_id );
```
Is there a way to avoid the extra query reloading the model?

ayushpahwa replied:
I added this due to some flakiness I was seeing. I triaged it and there's only one case where the reload is needed. When the locations array is cleared, I noticed the `$zone->get_locations()` method was returning the old data but not when the update was non-empty. I dug a bit and I suspect the issue might be in `WC_Data::apply_changes()`. The `array_replace_recursive()` is not correctly clearing the array. I checked with the v2/v3 impl and they seem to be always reloading. For now I have only added the reload when the locations are cleared. Please let me know if this makes sense.

ObliviousHarmony commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/ShippingZoneMethod/Controller.php:
```diff
@@ -51,7 +59,8 @@ class Controller extends AbstractController {
 	final public function init( ShippingMethodSchema $method_schema ) {
-		$this->method_schema = $method_schema;
+		$this->method_schema    = $method_schema;
+		$this->shipping_service = new ShippingService();
```
The DI container uses reflection to pass arguments to `init` methods based on the type hints. Rather than instantiating `ShippingService`, it should be injected.

ayushpahwa replied:
Thanks for catching this, I have updated it now!

ObliviousHarmony commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/ShippingZones/Controller.php:
```diff
@@ -54,7 +62,8 @@ class Controller extends AbstractController {
 	final public function init( ShippingZoneSchema $zone_schema ) {
-		$this->item_schema = $zone_schema;
+		$this->item_schema      = $zone_schema;
+		$this->shipping_service = new ShippingService();
```
This should be injected too.

ayushpahwa replied:
Updated

ObliviousHarmony commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/ShippingZoneMethod/ShippingZoneMethodService.php:
The name `ShippingService` in the `Internal\Shipping` namespace feels like it implies that it's a service that has something to do with shipping. Do you have any thoughts on an alternative name that is more precise about what this class does?

ObliviousHarmony replied:
`ShippingZoneService`?

ayushpahwa replied:
Yeah, this makes sense. I have broken down the services to reflect their names.

ObliviousHarmony commented on plugins/woocommerce/src/Internal/Shipping/ShippingService.php:
```diff
@@ -0,0 +1,289 @@
+		if ( 0 === $zone->get_id() ) {
+			if ( isset( $params['name'] ) && ! is_null( $params['name'] ) ) {
+				return new WP_Error(
+					'woocommerce_rest_cannot_edit_zone',
```
These error codes feel a little too specific for a general-purpose service class!

ObliviousHarmony commented on plugins/woocommerce/src/Internal/Shipping/ShippingService.php:
```diff
@@ -0,0 +1,289 @@
+	 * Update zone details from REST API request.
+	 *
+	 * @param WC_Shipping_Zone $zone zone to be updated.
+	 * @param array            $params Request parameters.
```
We should make sure our service isn't explicitly tied to the REST API. Would you mind expanding this doc block to describe the specific options that it accepts?

ayushpahwa replied:
Makes sense, I went through the other comments too and removed the redundant ones.

ObliviousHarmony commented on plugins/woocommerce/src/Internal/Shipping/ShippingService.php:
```diff
@@ -0,0 +1,289 @@
+	 * @param array            $params Request parameters.
+	 * @return true|WP_Error True on success, WP_Error on failure.
+	 */
+	public function update_shipping_zone( $zone, $params ) {
```
How would you feel about using `wp_parse_args()` to normalize the input and conveniently document all of the available parameters?

ayushpahwa replied:
Thanks for the pointer, I have added it to the update functions.

ObliviousHarmony reviewed (changes requested):
Hey @ayushpahwa,

First of all, thank you _very_ much for working on improving this code. I'm happy with the direction, however, the `ShippingService` feels like a collection of methods that are single endpoints in a REST API. They include REST parameter parsing and `WP_Error` messages built for REST responses.

While I'm not opposed to a separate shipping zone service, in practice, I think the service you've built here should have its functionality moved into the related controllers. It just doesn't seem like it provides value outside of that context but is treated as an application-wide service.

ObliviousHarmony commented on plugins/woocommerce/src/Internal/Shipping/ShippingService.php:
```diff
@@ -0,0 +1,289 @@
+	 * @param array            $params Request parameters.
+	 * @return true|WP_Error True on success, WP_Error on failure.
+	 */
+	public function update_shipping_zone( $zone, $params ) {
```
I wonder if this kind of validation would better fit the `validate_callback` for each arg? As far as I can tell, none of the validation here executes any database queries and there's an endpoint `validate_callback` that lets you check the request where you could do database query validation and loading.

ayushpahwa replied:
This looks like a good approach but would mean a bunch of refactor too. Would it be okay to pick this up post CIAB MVP? Anyways the V4 apis are still behind feature flag and internal so should be okay to make changes. What do you think?

ayushpahwa committed:
Merge branch 'trunk' into fix/woo13-84/address-maintainability-issues-with-v4-implementation

ayushpahwa committed:
update: break the service into 2 parts for cohesion

ayushpahwa committed:
update: fix controller references to new service

ayushpahwa committed:
test: split the service tests based on controller

ayushpahwa committed:
Merge branch 'trunk' into fix/woo13-84/address-maintainability-issues-with-v4-implementation

ayushpahwa committed:
delete: remove redundant set enabled method

ayushpahwa committed:
update: send updated shipping method value from service to controller

ayushpahwa committed:
update: send updated shipping zone value from service to controller

ayushpahwa committed:
Merge branch 'trunk' into fix/woo13-84/address-maintainability-issues-with-v4-implementation

ayushpahwa committed:
test: update the asserts to match the new func schema

The funcs earlier returned true if the update succeeded but now they
return the updated object. Hence the tests need to be updated

ayushpahwa committed:
update: remove redundant reload call for shipping method

ayushpahwa committed:
update: pass the dependencies as args in init

Following the DI principles, the deps for shipping controllers should be
passed as args of the contructor rather then calling their constructors.

ayushpahwa committed:
test: update setup to pass the new dependency

ayushpahwa committed:
Merge branch 'trunk' into fix/woo13-84/address-maintainability-issues-with-v4-implementation

ayushpahwa committed:
chore: remove unused params

ayushpahwa committed:
update: add workaround for wc data issue

ayushpahwa committed:
chore: fix incorrect location clearing logic

ayushpahwa committed:
test: update assertion to expect the new func schema

ayushpahwa committed:
update: add hook for backward compatibility with v2/v3

ayushpahwa committed:
chore: move service tests near the controller tests

ayushpahwa committed:
update: use parse args to set defaults of the update funcs

ayushpahwa commented on plugins/woocommerce/src/Internal/RestApi/Routes/V4/ShippingZoneMethod/ShippingZoneMethodService.php:
```diff
@@ -0,0 +1,156 @@
+			array( 'instance_id' => $instance_id ),
+			$formats,
+			array( '%d' )
+		);
```
Thanks, this is now fixed.

ayushpahwa committed:
chore: update doc block comments

ayushpahwa committed:
Merge branch 'trunk' into fix/woo13-84/address-maintainability-issues-with-v4-implementation

ayushpahwa committed:
chore: remove redundant comments

ayushpahwa committed:
chore: fix lint issues

ayushpahwa committed:
Merge branch 'trunk' into fix/woo13-84/address-maintainability-issues-with-v4-implementation

ayushpahwa commented:
Thanks for the insightful comments @ObliviousHarmony and @prettyboymp! I was trying to follow what the payments service was doing but it makes sense to keep the service near the controller. I have split the service and moved them with their respective controllers. Please let me know if this approach makes sense.

ObliviousHarmony commented on plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version4/ShippingZoneMethod/class-wc-rest-shipping-zone-method-v4-controller-tests.php:
I'm not going to make a big deal out of it because it was already done, however, none of the `Version4` tests are in the right place! All code in the PSR-4 namespace (`src/`) belongs in the `tests/php/src` directory with the same folder structure and namespace path.

ObliviousHarmony reviewed (approved):
Thanks @ayushpahwa,

Given that this is an experimental API, I'm comfortable at this point approving it. I really appreciate you taking the time here to improve your REST endpoint!

I _do_ have a few suggestions and requests for the future:

- To be honest, all of the tests for the V4 API are in the wrong place! The `tests/php` directory has two places for tests to go. These mirror the file path of the source file under test. Anything in `include` goes in `tests/php/include` and everything in `src` goes in `tests/php/src`.
- The intention behind dependency injection is to make it easy to abstract a class away from the others that it interacts with. In practical terms, tests don't need to receive actual classes and can receive mocks. This lets you isolate the code under test, for example, controller classes can make sure _they_ work without making sure that the service they consume does too. As an added bonus, this lets you speed up tests by avoiding the side effects of calling the other class such as database access and the need for global configuration state.

ayushpahwa committed:
Merge branch 'trunk' into fix/woo13-84/address-maintainability-issues-with-v4-implementation
